### PR TITLE
Make AsyncCall handle interceptors that throws an IOException

### LINKED ISSUE #1801: AsyncCall NPE's if an interceptor throws before the call is attempted.
AsyncCall throws an uncaught NullPointerException [here](https://github.com/square/okhttp/blob/master/okhttp/src/main/java/com/squareup/okhttp/Call.java#L175) because engine is null, having never been initialized, if the call was never attempted.  For example, this can happen if an ("application-level") interceptor aborted the call by throwing an IOException before the call was ever attempted.

I suggest that it could simply pass a null request to the callback in this case instead.

### PR DESCRIPTION
**Pull request by ansman**
This closes #1801

### CODE REVIEW

File: okhttp/src/main/java/com/squareup/okhttp/Call.java [hunk 1]
```diff
@@ -172,7 +172,8 @@ Call get() {
           // Do not signal the callback twice!
           logger.log(Level.INFO, "Callback failure for " + toLoggableString(), e);
         } else {
-          responseCallback.onFailure(engine.getRequest(), e);
+          Request request = engine == null ? null : engine.getRequest();
+          responseCallback.onFailure(request, e);
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/Call.java]: We've already called `onFailure()` when we get here.

Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/Call.java]: hang on, I'm reading this wrong.

File: okhttp/src/main/java/com/squareup/okhttp/Call.java [hunk 2]
```diff
@@ -172,7 +172,8 @@ Call get() {
           // Do not signal the callback twice!
           logger.log(Level.INFO, "Callback failure for " + toLoggableString(), e);
         } else {
-          responseCallback.onFailure(engine.getRequest(), e);
+          Request request = engine == null ? null : engine.getRequest();
```
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/Call.java]: use `originalRequest` ?

Comment (ansman) [inline okhttp/src/main/java/com/squareup/okhttp/Call.java]: Doh, of course. I'll change so it uses original request if the engine is `null`

Comment (ansman) [inline okhttp/src/main/java/com/squareup/okhttp/Call.java]: Implemented as requested, let me know if I should squash the commits or not.

File: okhttp-tests/src/test/java/com/squareup/okhttp/CallTest.java [hunk 3]
```diff
@@ -697,6 +697,19 @@ private void postBodyRetransmittedAfterAuthorizationFail(String body) throws Exc
     }
   }
 
+  @Test public void asyncCallEngineInitialized () throws Exception {
+    OkHttpClient client = new OkHttpClient();
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/CallTest.java]: For what it's worth, this shadows another variable called `client`.

File: okhttp-tests/src/test/java/com/squareup/okhttp/CallTest.java [hunk 4]
```diff
@@ -697,6 +697,19 @@ private void postBodyRetransmittedAfterAuthorizationFail(String body) throws Exc
     }
   }
 
+  @Test public void asyncCallEngineInitialized () throws Exception {
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/CallTest.java]: ultranit: no ` ` before `()`.

File: okhttp-tests/src/test/java/com/squareup/okhttp/CallTest.java [hunk 5]
```diff
@@ -697,6 +697,19 @@ private void postBodyRetransmittedAfterAuthorizationFail(String body) throws Exc
...
+    OkHttpClient client = new OkHttpClient();
+    client.interceptors().add(new Interceptor() {
+      @Override public Response intercept(Chain chain) throws IOException {
+        throw new IOException();
+      }
+    });
+    Request request = new Request.Builder().url(server.url("/")).build();
+    client.newCall(request).enqueue(callback);
+    RecordedResponse response = callback.await(request.httpUrl());
+    assertEquals(request, response.request);
```
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/CallTest.java]: Nice test!

**Comment by swankjesse** [comment]
LGTM, just two tiny nitpicks. Sorry for sitting on this; GitHub doesn't email me when you push new code on a PR.

**Comment by swankjesse** [comment]
Also- please http://rebaseandsqua.sh.

**Comment by swankjesse** [comment]
Please complete Square's contributors agreement.
https://docs.google.com/forms/d/13WR8m5uZ2nAkJH41k7GdVBXAAbzDk00vxtEYjd6Imzg/viewform?formkey=dDViT2xzUHAwRkI3X3k5Z0lQM091OGc6MQ&ndplr=1

**Comment by ansman** [comment]
@swankjesse 
 * A comment to reference #1801 was made before the test
 * Extraneous space was removed
 * Client variable shadowing was fixed
 * Rebased and squased

**Comment by swankjesse** [comment]
Sweet. Thank you!

**Comment by GrahamFirst** [comment]
Awesome to see this fix!  Any idea when it will be released?

**Comment by swankjesse** [comment]
@GrahamFirst I'm hoping to finish 2.6 in 2015.

**Pull request merged**